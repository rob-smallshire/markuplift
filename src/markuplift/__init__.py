"""MarkupLift - A configurable XML and HTML formatter.

MarkupLift provides flexible formatting of XML and HTML documents with configurable
predicates for block vs inline elements, whitespace handling, and custom text content
formatters. The package uses a factory pattern with ElementPredicateFactory functions
for optimal performance.

Main Classes:
    Formatter: High-level API for formatting documents
    DocumentFormatter: Low-level document-specific formatter

Example:
    from markuplift import Formatter
    from markuplift.predicates import tag_equals, html_block_elements

    formatter = Formatter(
        block_predicate_factory=html_block_elements(),
        inline_predicate_factory=tag_equals("span")
    )
    result = formatter.format_str("<div><span>content</span></div>")
"""

from .formatter import Formatter
from .document_formatter import DocumentFormatter
from .types import ElementPredicate, ElementPredicateFactory
from .annotation import INLINE_TYPE_ANNOTATION as INLINE_TYPE
from .annotation import BLOCK_TYPE_ANNOTATION as BLOCK_TYPE
from .predicates import (
    all_of,
    any_of,
    attribute_count_between,
    attribute_count_max,
    attribute_count_min,
    attribute_equals,
    has_attribute,
    has_child_elements,
    has_mixed_content,
    has_no_significant_content,
    has_significant_content,
    html_block_elements,
    html_inline_elements,
    html_metadata_elements,
    html_void_elements,
    html_whitespace_significant_elements,
    is_comment,
    is_element,
    is_processing_instruction,
    matches_xpath,
    not_matching,
    tag_equals,
    tag_in,
    PredicateError,
)

__all__ = [
    "DocumentFormatter",
    "Formatter",
    "ElementPredicate",
    "ElementPredicateFactory",
    "PredicateError",
    "INLINE_TYPE",
    "BLOCK_TYPE",
    "all_of",
    "any_of",
    "attribute_count_between",
    "attribute_count_max",
    "attribute_count_min",
    "attribute_equals",
    "has_attribute",
    "has_child_elements",
    "has_mixed_content",
    "has_no_significant_content",
    "has_significant_content",
    "html_block_elements",
    "html_inline_elements",
    "html_metadata_elements",
    "html_void_elements",
    "html_whitespace_significant_elements",
    "is_comment",
    "is_element",
    "is_processing_instruction",
    "matches_xpath",
    "not_matching",
    "tag_equals",
    "tag_in",
]

from collections import namedtuple

Version = namedtuple("Version", ["major", "minor", "patch"])
__version__ = "2.1.0"
__version_info__ = Version(*(__version__.split(".")))
