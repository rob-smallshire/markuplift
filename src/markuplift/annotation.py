"""Element annotation system for XML/HTML formatting.

This module provides a comprehensive annotation system that attaches metadata
to XML elements during the formatting process. Annotations are used to store
information about element types (block/inline), whitespace handling, indentation
levels, and other formatting-related properties.

The annotation system works with ElementPredicate functions to efficiently
categorize and process elements based on their characteristics and context.

Key Components:
    Annotations: Storage class for element metadata
    Element type annotation functions (block/inline classification)
    Whitespace processing annotation functions
    Indentation level calculation functions
    Text transformation functions
"""

from typing import Any, Callable
from functools import partial
from enum import Enum

from lxml import etree

# Import ElementPredicate type alias
from markuplift.types import ElementPredicate

from markuplift.utilities import (
    is_in_mixed_content, parent_is_annotated_with, normalize_ws,
    siblings,
)

LOGICAL_LEVEL_ANNOTATION_KEY = "logical_level"

PHYSICAL_LEVEL_ANNOTATION_KEY = "physical_level"

XML_SPACE_DEFAULT = "default"
XML_SPACE_PRESERVE = "preserve"

WHITESPACE_ANNOTATION_KEY = "whitespace"
PRESERVE_WHITESPACE_ANNOTATION = "preserve"
NORMALIZE_WHITESPACE_ANNOTATION = "normalize"
STRIP_WHITESPACE_ANNOTATION = "strip"  # Strip implies normalize
STRICT_WHITESPACE_ANNOTATION = "strict" # No transformations at all
# If there is no value associated with WHITESPACE_ANNOTATION_KEY , the formatter can take action
# contextually, e.g. based on whether the element is block or inline.

TYPE_ANNOTATION_KEY = "type"
BLOCK_TYPE_ANNOTATION = "block"
INLINE_TYPE_ANNOTATION = "inline"
# If there is no value associated with TYPE_ANNOTATION_KEY , the formatter can take action
# contextually, e.g. based on whether the element is in mixed content or not.
BLOCK_TYPES = {None, BLOCK_TYPE_ANNOTATION, INLINE_TYPE_ANNOTATION}

class Annotations:

    def __init__(self):
        self._annotations: dict[etree._Element, dict[str, str]] = {}

    def annotate(self, element: etree._Element, attribute_name: str, attribute_value: Any):
        if element not in self._annotations:
            self._annotations[element] = {}
        self._annotations[element][attribute_name] = attribute_value

    def annotation(self, element: etree._Element, attribute_name: str, default: Any = None) ->Any:
        return self._annotations.get(element, {}).get(attribute_name, default)


class AnnotationConflictMode(Enum):
    RAISE = "raise"
    SKIP = "skip"
    OVERWRITE = "overwrite"


class AnnotationConflictError(Exception):
    pass


def annotate_explicit_block_elements(
    root: etree._Element,
    annotations: Annotations,
    block_predicate: ElementPredicate,
):
    annotate_matches(root, annotations, block_predicate, TYPE_ANNOTATION_KEY, BLOCK_TYPE_ANNOTATION, conflict_mode=AnnotationConflictMode.RAISE)


def annotate_explicit_inline_elements(
    root: etree._Element,
    annotations: Annotations,
    inline_predicate: ElementPredicate,
):
    annotate_matches(root, annotations, inline_predicate, TYPE_ANNOTATION_KEY, INLINE_TYPE_ANNOTATION, conflict_mode=AnnotationConflictMode.RAISE)


def annotate_elements_in_mixed_content_as_inline(
    root: etree._Element,
    annotations: Annotations,
):
    annotate_matches(root, annotations, is_in_mixed_content, TYPE_ANNOTATION_KEY, INLINE_TYPE_ANNOTATION, conflict_mode=AnnotationConflictMode.SKIP)


def annotate_inline_descendants_as_inline(
    root: etree._Element,
    annotations: Annotations,
):
    """The inline nature of an element is inherited by its descendants unless they are already annotated with a type."""
    annotate_matches(
        root,
        annotations,
        partial(parent_is_annotated_with, annotations=annotations, annotation_key=TYPE_ANNOTATION_KEY, annotation_value=INLINE_TYPE_ANNOTATION),
        TYPE_ANNOTATION_KEY,
        INLINE_TYPE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.SKIP,
    )


def annotate_unmixed_block_descendants_as_block(
    root: etree._Element,
    annotations: Annotations,
):
    """The block nature of an element is inherited by its descendants iff they are element-only (i.e. are not mixed with significant text)
    and none of the siblings are inline.
    """
    # We need to combine two predicates: parent is annotated as block, and element is not in mixed content
    annotate_matches(
        root,
        annotations,
        lambda e: (
            parent_is_annotated_with(e, annotations, TYPE_ANNOTATION_KEY, BLOCK_TYPE_ANNOTATION)
            and
            (not is_in_mixed_content(e))
            and not any(
                annotations.annotation(sibling, TYPE_ANNOTATION_KEY) == INLINE_TYPE_ANNOTATION
                for sibling in siblings(e)
            )
        ),
        TYPE_ANNOTATION_KEY,
        BLOCK_TYPE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.SKIP,
    )


def annotate_xml_space(
    root: etree._Element,
    annotations: Annotations,
):
    """Annotate elements with xml:space="preserve" with whitespace='preserve'.

    This attribute propagates to descendants unless it is overridden by xml:space="default", though
    in practice we'll consider any value other than "preserve" to be "default" and stop propagation.

    We mark elements affected by xml:space="preserve" with a 'whitespace' annotation with value
    'preserve'.
    """
    annotate_matches(
        root,
        annotations,
        lambda e: (
            e.get("{http://www.w3.org/XML/1998/namespace}space") == XML_SPACE_PRESERVE
            or
            (
                parent_is_annotated_with(e, annotations, WHITESPACE_ANNOTATION_KEY, STRICT_WHITESPACE_ANNOTATION)
                and
                e.get("{http://www.w3.org/XML/1998/namespace}space") != XML_SPACE_DEFAULT
            )
        ),
        WHITESPACE_ANNOTATION_KEY,
        STRICT_WHITESPACE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.OVERWRITE,
    )


def annotate_explicit_whitespace_preserving_elements(
    root: etree._Element,
    annotations: Annotations,
    predicate: ElementPredicate,
):
    annotate_matches(
        root,
        annotations,
        predicate,
        WHITESPACE_ANNOTATION_KEY,
        PRESERVE_WHITESPACE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.OVERWRITE
    )


def annotate_whitespace_preserving_descendants_as_whitespace_preserving(
    root: etree._Element,
    annotations: Annotations,
):
    """The whitespace-preserving nature of an element is inherited by its descendants unless they are already annotated with whitespace."""
    annotate_matches(
        root,
        annotations,
        partial(parent_is_annotated_with, annotations=annotations, annotation_key=WHITESPACE_ANNOTATION_KEY, annotation_value=PRESERVE_WHITESPACE_ANNOTATION),
        WHITESPACE_ANNOTATION_KEY,
        PRESERVE_WHITESPACE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.SKIP,
    )


def annotate_explicit_whitespace_normalizing_elements(
    root: etree._Element,
    annotations: Annotations,
    predicate: ElementPredicate,
):
    annotate_matches(
        root,
        annotations,
        predicate,
        WHITESPACE_ANNOTATION_KEY,
        NORMALIZE_WHITESPACE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.OVERWRITE
    )


def annotate_explicit_stripped_elements(
    root: etree._Element,
    annotations: Annotations,
    predicate: ElementPredicate,
):
    annotate_matches(
        root,
        annotations,
        predicate,
        WHITESPACE_ANNOTATION_KEY,
        STRIP_WHITESPACE_ANNOTATION,
        conflict_mode=AnnotationConflictMode.OVERWRITE
    )


def annotate_untyped_elements_as_default(
    root: etree._Element,
    annotations: Annotations,
    default_type: str,
):
    if default_type not in BLOCK_TYPES:
        raise ValueError(f"default_type must be one of {BLOCK_TYPES}")
    annotate_matches(
        root,
        annotations,
        lambda e: annotations.annotation(e, TYPE_ANNOTATION_KEY) is None,
        TYPE_ANNOTATION_KEY,
        default_type,
        conflict_mode=AnnotationConflictMode.SKIP,
    )


def annotate_logical_level(
    root: etree._Element,
    annotations: Annotations,
):
    annotations.annotate(root, LOGICAL_LEVEL_ANNOTATION_KEY, 0)
    for elem in root.iter():
        parent = elem.getparent()
        if parent is not None:
            parent_level = annotations.annotation(parent, LOGICAL_LEVEL_ANNOTATION_KEY)
            if parent_level is not None:
                annotations.annotate(elem, LOGICAL_LEVEL_ANNOTATION_KEY, parent_level + 1)


def annotate_physical_level(
    root: etree._Element,
    annotations: Annotations,
):
    annotations.annotate(root, PHYSICAL_LEVEL_ANNOTATION_KEY, 0)
    for elem in root.iter():
        parent = elem.getparent()
        if parent is not None:
            parent_level = annotations.annotation(parent, PHYSICAL_LEVEL_ANNOTATION_KEY)
            if parent_level is not None:
                parent_type = annotations.annotation(parent, TYPE_ANNOTATION_KEY)
                if parent_type == INLINE_TYPE_ANNOTATION:
                    physical_level = parent_level
                elif parent_type == BLOCK_TYPE_ANNOTATION:
                    physical_level = parent_level + 1
                else:
                    assert parent_type is None
                    physical_level = parent_level  # Preserve existing structure
                annotations.annotate(elem, PHYSICAL_LEVEL_ANNOTATION_KEY, physical_level)


def annotate_text_transforms(
    root: etree._Element,
    annotations: Annotations,
    one_indent: str,
):
    # The text of an element comes between the element's start tag and the start tag of its first
    # child (or its end tag if it has no children). This function will emplace an attribute which
    # describes how this text should be transformed based on existing annotations on the element and
    # its first child (if any).

    for elem in root.iter():
        text_transforms: list[Callable[[str], str]] = []
        whitespace = annotations.annotation(elem, WHITESPACE_ANNOTATION_KEY)
        first_child = next(iter(elem), None)
        first_child_type = (
            annotations.annotation(first_child, TYPE_ANNOTATION_KEY)
            if (first_child is not None) else
            None
        )

        if whitespace not in {PRESERVE_WHITESPACE_ANNOTATION, STRICT_WHITESPACE_ANNOTATION}:
            if whitespace in {NORMALIZE_WHITESPACE_ANNOTATION, STRIP_WHITESPACE_ANNOTATION}:
                text_transforms.append(normalize_ws)
                if whitespace == STRIP_WHITESPACE_ANNOTATION:
                    text_transforms.append(str.lstrip)
            if first_child_type == BLOCK_TYPE_ANNOTATION:
                child_physical_level = annotations.annotation(
                    first_child, PHYSICAL_LEVEL_ANNOTATION_KEY, 0
                )
                text_transform = partial(
                    transform_text_preceding_block, physical_level=child_physical_level,
                    one_indent=one_indent
                )
                text_transforms.append(text_transform)

            if first_child is None:
                if whitespace == STRIP_WHITESPACE_ANNOTATION:
                    text_transforms.append(str.rstrip)

        annotations.annotate(elem, "text_transforms", text_transforms)


def annotate_tail_transforms(root, annotations, one_indent):
    # The tail text of an element comes between the element's end tag and the start tag of its
    # next sibling (or its parent's end tag if it has no next sibling). This function will emplace
    # an attribute which describes how this tail text should be transformed based on existing
    # annotations on the element and its next sibling (if any).

    for elem in root.iter():
        tail_transforms: list[Callable[[str], str]] = []

        # Tail text exists within the parent element, so we consider the parent's whitespace annotation
        # when determining how to transform the tail text.
        parent = elem.getparent()
        parent_whitespace = (
            annotations.annotation(parent, WHITESPACE_ANNOTATION_KEY)
            if (parent is not None) else
            None
        )

        parent_physical_level = (
            annotations.annotation(parent, PHYSICAL_LEVEL_ANNOTATION_KEY, 0)
            if (parent is not None) else
            0
        )

        next_sibling = elem.getnext()
        next_sibling_type = (
            annotations.annotation(next_sibling, TYPE_ANNOTATION_KEY)
            if (next_sibling is not None) else
            None
        )

        # We also don't need to consider the element's own whitespace annotation, since that only
        # affects the element's text, not its tail. However, we do consider the element's
        # type, since that can affect how we treat the tail text. Tail text can also precede a block
        # element, so we need to consider any following sibling elements as well.
        elem_type = annotations.annotation(elem, TYPE_ANNOTATION_KEY)
        if parent_whitespace not in {PRESERVE_WHITESPACE_ANNOTATION, STRICT_WHITESPACE_ANNOTATION}:
            if parent_whitespace in {NORMALIZE_WHITESPACE_ANNOTATION, STRIP_WHITESPACE_ANNOTATION}:
                tail_transforms.append(normalize_ws)
                if next_sibling is None:
                    if parent_whitespace == STRIP_WHITESPACE_ANNOTATION:
                        tail_transforms.append(str.rstrip)
            if elem_type == BLOCK_TYPE_ANNOTATION:
                if next_sibling_type in {BLOCK_TYPE_ANNOTATION}:
                    text_transform = partial(
                        transform_text_following_block, physical_level=parent_physical_level,
                        one_indent=one_indent
                    )
                    tail_transforms.append(text_transform)
                elif next_sibling_type == INLINE_TYPE_ANNOTATION:
                    text_transform = partial(
                        transform_text_following_block_preceding_inline, physical_level=parent_physical_level,
                    )
                    tail_transforms.append(text_transform)  # Just add a newline before the text
                else:
                    if parent is not None:
                        text_transform = partial(
                            transform_text_following_block, physical_level=parent_physical_level,
                            one_indent=one_indent
                        )
                        tail_transforms.append(text_transform)
                    else:
                        assert parent is None
                        # If the element is at logical level 0, it is the root element, so we
                        # are not allowed to have any tail text at all.
                        tail_transforms.append(lambda s: "")

            if next_sibling is not None:
                if next_sibling_type == BLOCK_TYPE_ANNOTATION:
                    sibling_physical_level = annotations.annotation(
                        next_sibling, PHYSICAL_LEVEL_ANNOTATION_KEY, 0
                    )
                    text_transform = partial(
                        transform_text_preceding_block, physical_level=sibling_physical_level,
                        one_indent=one_indent
                    )
                    tail_transforms.append(text_transform)

        annotations.annotate(elem, "tail_transforms", tail_transforms)



def transform_text_preceding_block(text: str, physical_level: int, one_indent: str) -> str:
    # 1. Strip trailing whitespace
    text = text.rstrip()
    # 2. Add newline and indentation after the text to separate it from the block child
    indent = one_indent * physical_level
    text = text + "\n" + indent
    return text


def transform_text_following_block(text: str, physical_level: int, one_indent: str) -> str:
    # 1. Strip leading whitespace
    text = text.lstrip()
    # 2. Add newline and indentation before the text to separate it from the block element
    indent = one_indent * physical_level
    text = "\n" + indent + text
    return text


def transform_text_following_block_preceding_inline(text: str, physical_level: int) -> str:
    # If the leading run of whitespace does not contain a newline, we add one
    leading_ws = len(text) - len(text.lstrip())
    if leading_ws > 0:
        if "\n" not in text[:leading_ws]:
            text = "\n" + text
    return text


def annotate_matches(
    tree: etree._Element,
    annotations: Annotations,
    predicate: ElementPredicate,
    annotation_key: str,
    annotation_value: str,
    *,
    conflict_mode: AnnotationConflictMode = AnnotationConflictMode.RAISE,
):
    """Annotate elements in the tree that match the predicate with the given annotation.

    Args:
        tree: The XML tree to annotate.
        annotations: The Annotations object to store annotations.
        predicate: A function that takes an element and returns True if it should be annotated.
        annotation_key: The key for the annotation.
        annotation_value: The value for the annotation.
        conflict_mode: Determines how to handle conflicts with existing annotations.
    """
    for elem in tree.iter():
        if predicate(elem):
            existing_type = annotations.annotation(elem, annotation_key)
            if existing_type is not None:
                if conflict_mode == AnnotationConflictMode.RAISE:
                    raise AnnotationConflictError(
                        f"Element <{elem.tag}> was previously marked as {existing_type}, cannot also mark as {annotation_value}"
                    )
                elif conflict_mode == AnnotationConflictMode.SKIP:
                    continue
                elif conflict_mode == AnnotationConflictMode.OVERWRITE:
                    pass  # Overwrite annotation
            annotations.annotate(elem, annotation_key, annotation_value)

