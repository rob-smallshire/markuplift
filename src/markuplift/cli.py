"""Command-line interface for MarkupLift.

This module provides the CLI for MarkupLift, allowing users to format XML and HTML
documents from the command line with configurable options for block/inline elements,
whitespace handling, and external text formatters.

The CLI uses ElementPredicateFactory functions from the predicates module to create
optimized formatting rules based on XPath expressions provided by the user.
"""

import subprocess

import click
from lxml import etree

from markuplift.formatter import Formatter
from markuplift.predicates import matches_xpath, PredicateError



@click.group()
@click.version_option()
def cli():
    """MarkupLift - A configurable XML and HTML formatter.

    MarkupLift provides flexible formatting of XML and HTML documents with
    configurable predicates for block vs inline elements, whitespace handling,
    and custom text content formatters.
    """
    pass


@cli.command()
@click.argument("input_file", type=click.File("r"), default="-")
@click.option("--output", "-o", type=click.File("w"), default="-", help="Output file (default: stdout)")
@click.option("--block", multiple=True, help="XPath expression for block elements (can be used multiple times)")
@click.option("--inline", multiple=True, help="XPath expression for inline elements (can be used multiple times)")
@click.option(
    "--normalize-whitespace", multiple=True, help="XPath expression for elements that should have normalized whitespace"
)
@click.option(
    "--preserve-whitespace", multiple=True, help="XPath expression for elements that should preserve whitespace"
)
@click.option(
    "--strip-whitespace", multiple=True, help="XPath expression for elements that should have whitespace stripped"
)
@click.option("--wrap-attributes", multiple=True, help="XPath expression for elements that should wrap attributes")
@click.option(
    "--text-formatter",
    nargs=2,
    multiple=True,
    metavar="XPATH COMMAND",
    help="Apply external formatter to text content. XPATH selects elements, COMMAND is the external program.",
)
@click.option("--indent-size", type=int, default=2, help="Number of spaces for each indentation level (default: 2)")
@click.option(
    "--default-type",
    type=click.Choice(["block", "inline"]),
    default="block",
    help="Default element type for unclassified elements (default: block)",
)
@click.option(
    "--xml-declaration/--no-xml-declaration", default=False, help="Include XML declaration in output (default: no)"
)
@click.option("--doctype", help="Custom DOCTYPE declaration to include")
def format(
    input_file,
    output,
    block,
    inline,
    normalize_whitespace,
    preserve_whitespace,
    strip_whitespace,
    wrap_attributes,
    text_formatter,
    indent_size,
    default_type,
    xml_declaration,
    doctype,
):
    """Format XML or HTML documents with configurable formatting rules.

    INPUT_FILE can be a file path or '-' for stdin (default).

    Examples:

    \b
    # Format from file to stdout
    markuplift format input.xml

    \b
    # Format with custom block/inline elements
    markuplift format input.html --block "//div | //section" --inline "//em | //strong"

    \b
    # Use external formatter for JavaScript in script tags
    markuplift format input.html --text-formatter "//script[@type='javascript']" "js-beautify"

    \b
    # Format from stdin to file
    cat input.xml | markuplift format --output formatted.xml
    """
    try:
        # Read input content
        content = input_file.read()

        # Create combined predicate factories from XPath expressions
        def combine_factories(xpath_list):
            if not xpath_list:
                return None
            try:
                factories = [matches_xpath(xpath) for xpath in xpath_list]
            except PredicateError as e:
                raise click.ClickException(str(e))
            def combined_factory(root):
                predicates = [factory(root) for factory in factories]
                return lambda e: any(pred(e) for pred in predicates)
            return combined_factory

        # Create text formatter factories from external programs
        text_formatter_factories = {}
        for xpath_expr, command in text_formatter:
            try:
                factory = matches_xpath(xpath_expr)
            except PredicateError as e:
                raise click.ClickException(str(e))
            def create_formatter(cmd=command):  # Capture command in closure
                def formatter_func(text, doc_formatter, physical_level):
                    if not text.strip():
                        return text
                    try:
                        cmd_parts = cmd.split()
                        result = subprocess.run(cmd_parts, input=text, text=True, capture_output=True, timeout=30)
                        if result.returncode != 0:
                            click.echo(f"Warning: External formatter '{cmd}' failed: {result.stderr}", err=True)
                            return text
                        return result.stdout
                    except subprocess.TimeoutExpired:
                        click.echo(f"Warning: External formatter '{cmd}' timed out", err=True)
                        return text
                    except FileNotFoundError:
                        click.echo(f"Warning: External formatter command '{cmd.split()[0]}' not found", err=True)
                        return text
                    except Exception as e:
                        click.echo(f"Warning: External formatter '{cmd}' error: {e}", err=True)
                        return text
                return formatter_func
            text_formatter_factories[factory] = create_formatter()

        # Create formatter with factory functions - much cleaner!
        formatter = Formatter(
            block_predicate_factory=combine_factories(block),
            inline_predicate_factory=combine_factories(inline),
            normalize_whitespace_predicate_factory=combine_factories(normalize_whitespace),
            strip_whitespace_predicate_factory=combine_factories(strip_whitespace),
            preserve_whitespace_predicate_factory=combine_factories(preserve_whitespace),
            wrap_attributes_predicate_factory=combine_factories(wrap_attributes),
            text_content_formatters=text_formatter_factories,
            indent_size=indent_size,
            default_type=default_type,
        )

        # Format the content - Formatter handles parsing and optimization internally
        formatted = formatter.format_str(content, doctype=doctype, xml_declaration=xml_declaration)

        # Write output
        output.write(formatted)

    except etree.XMLSyntaxError as e:
        raise click.ClickException(f"XML parsing error: {e}")
    except Exception as e:
        raise click.ClickException(f"Formatting error: {e}")


if __name__ == "__main__":
    cli()
