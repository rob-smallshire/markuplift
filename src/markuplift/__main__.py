from .cli import cli
cli()
import re
from dataclasses import dataclass, field
from typing import List, Optional
from lxml import html

@dataclass
class FormatterConfig:
    block_xpaths: List[str] = field(default_factory=lambda: [
        "//div", "//p", "//section", "//article",
        "//h1", "//h2", "//h3", "//ul", "//ol", "//li"
    ])
    inline_xpaths: List[str] = field(default_factory=lambda: [
        "//span", "//a", "//em", "//strong", "//b", "//i"
    ])
    self_closing_tags: List[str] = field(default_factory=lambda: [
        "br", "img", "hr", "meta", "link", "input"
    ])
    preserve_whitespace_xpaths: List[str] = field(default_factory=lambda: [
        "//pre", "//code", "//textarea"
    ])
    wrap_attributes_xpaths: List[str] = field(default_factory=list)
    wrap_attributes_threshold: int = 3
    max_line_length: int = 80
    indent: str = "  "


class HTMLFormatter:
    def __init__(self, config: Optional[FormatterConfig] = None):
        self.config = config or FormatterConfig()

    def format_html(self, html_string: str) -> str:
        parser = html.HTMLParser(remove_blank_text=True)
        tree = html.fromstring(html_string, parser=parser)

        # Precompute element sets for efficiency
        block_elems = self._collect_elements(tree, self.config.block_xpaths)
        inline_elems = self._collect_elements(tree, self.config.inline_xpaths)
        preserve_elems = self._collect_elements(tree, self.config.preserve_whitespace_xpaths)
        wrap_attr_elems = self._collect_elements(tree, self.config.wrap_attributes_xpaths)
        self_closing_tags = set(self.config.self_closing_tags)

        def normalize_ws(text):
            return re.sub(r'\s+', ' ', text).strip()

        def classify(elem):
            if elem in block_elems:
                return "block"
            elif elem in inline_elems:
                return "inline"
            else:
                return "other"

        def should_wrap_attributes(elem, is_block):
            if elem in wrap_attr_elems:
                return True
            if is_block and len(elem.attrib) >= self.config.wrap_attributes_threshold:
                return True
            tag_open = f"<{elem.tag}"
            attr_str = "".join(f' {k}="{v}"' for k, v in elem.attrib.items())
            if is_block and self.config.max_line_length and len(tag_open + attr_str + ">") > self.config.max_line_length:
                return True
            return False

        def format_attributes(elem, level, wrap):
            if not elem.attrib:
                return ""
            if not wrap:
                return "".join(f' {k}="{v}"' for k, v in elem.attrib.items())
            # Multi-line attributes
            parts = []
            for k, v in elem.attrib.items():
                parts.append(f"\n{self.config.indent * (level+1)}{k}=\"{v}\"")
            return "".join(parts) + "\n" + self.config.indent * (level+1)

        def recurse(elem, level=0, preserve_ws=False, parent_block=False):
            elem_type = classify(elem)
            is_self_closing = elem.tag in self_closing_tags
            preserve_ws = preserve_ws or elem in preserve_elems
            is_block = elem_type == "block"
            is_inline = elem_type == "inline"

            new_level = level + 1 if is_block else level
            parent_block = parent_block or is_block

            wrap_attrs = should_wrap_attributes(elem, is_block)

            # Start tag
            if is_block:
                result = "\n" + self.config.indent * level + f"<{elem.tag}"
            elif parent_block:
                result = f"<{elem.tag}"
            else:
                result = "\n" + self.config.indent * level + f"<{elem.tag}"

            # Attributes
            result += format_attributes(elem, level, wrap_attrs)

            if is_self_closing:
                result += "/>"
                return result

            result += ">"

            # Children or text
            children = list(elem)
            text = elem.text or ""
            if text:
                text_to_add = text if preserve_ws else normalize_ws(text)
                if text_to_add:
                    if is_block:
                        result += "\n" + self.config.indent * new_level + text_to_add
                    else:
                        result += text_to_add

            for child in children:
                result += recurse(child, new_level, preserve_ws, parent_block=is_block or parent_block)

            # Closing tag
            if is_block and children:
                result += "\n" + self.config.indent * level
            result += f"</{elem.tag}>"

            # Tail text
            if elem.tail:
                tail_to_add = elem.tail if preserve_ws else normalize_ws(elem.tail)
                if tail_to_add:
                    if parent_block:
                        result += "\n" + self.config.indent * level + tail_to_add
                    else:
                        result += tail_to_add

            return result

        return recurse(tree, 0).strip() + "\n"

    def _collect_elements(self, tree, xpaths: List[str]):
        elems = set()
        for xp in xpaths:
            elems.update(tree.xpath(xp))
        return elems


# Convenience API
def format_html(html_string: str, config: Optional[FormatterConfig] = None) -> str:
    return HTMLFormatter(config).format_html(html_string)
