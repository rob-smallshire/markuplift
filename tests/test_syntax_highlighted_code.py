"""Tests for formatting syntax-highlighted code blocks with preserved whitespace."""

from approvaltests import verify  # type: ignore
from markuplift import Html5Formatter


def test_syntax_highlighted_code_preserves_whitespace():
    """Test that syntax-highlighted code in <pre> preserves its whitespace structure."""
    html = """<body><article><section><pre class="code contain" id="code"><span></span><span id="line-1"><span class="k">def</span><span class="w"> </span><span class="nf">nth_root</span><span class="p">(</span><span class="n">radicand</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
</span>    <span class="fragment initial-highlighted-initial" data-fragment-group-id="single-return" id="line-2"><span class="k">return</span> <span class="n">radicand</span> <span class="o">**</span> <span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="n">n</span><span class="p">)</span>
</span><span id="line-3">
</span><span id="line-4">
</span><span id="line-5"><span class="k">def</span><span class="w"> </span><span class="nf">ordinal_suffix</span><span class="p">(</span><span class="n">value</span><span class="p">):</span>
</span>    <span id="line-6"><span class="n">s</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
</span>    <span id="line-7"><span class="k">if</span> <span class="n">s</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">'11'</span><span class="p">):</span>
</span>        <span class="fragment initial-highlighted-initial" data-fragment-group-id="multiple-return" id="line-8"><span class="k">return</span> <span class="s1">'th'</span>
</span>    <span id="line-9"><span class="k">elif</span> <span class="n">s</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">'12'</span><span class="p">):</span>
</span>        <span class="fragment initial-highlighted-initial" data-fragment-group-id="multiple-return" id="line-10"><span class="k">return</span> <span class="s1">'th'</span>
</span>    <span id="line-11"><span class="k">elif</span> <span class="n">s</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">'13'</span><span class="p">):</span>
</span>        <span class="fragment initial-highlighted-initial" data-fragment-group-id="multiple-return" id="line-12"><span class="k">return</span> <span class="s1">'th'</span>
</span></pre></section></article></body>"""

    # Html5Formatter now uses html_normalize_whitespace() by default which automatically
    # excludes whitespace-significant elements (<pre>, <style>, etc.) and their descendants
    formatter = Html5Formatter()

    result = formatter.format_str(html)
    verify(result)


def test_syntax_highlighted_code_with_attribute_reordering():
    """Test that syntax-highlighted code preserves whitespace while reordering attributes."""
    html = """<body><article><section><pre id="code" class="code contain"><span></span><span id="line-1"><span class="k">def</span><span class="w"> </span><span class="nf">nth_root</span><span class="p">(</span><span class="n">radicand</span><span class="p">):</span>
</span>    <span id="line-2" data-fragment-group-id="single-return" class="fragment initial-highlighted-initial"><span class="k">return</span> <span class="n">radicand</span>
</span></pre></section></article></body>"""

    # Html5Formatter now uses html_normalize_whitespace() by default which automatically
    # excludes whitespace-significant elements (<pre>, <style>, etc.) and their descendants
    formatter = Html5Formatter()

    result = formatter.format_str(html)
    verify(result)