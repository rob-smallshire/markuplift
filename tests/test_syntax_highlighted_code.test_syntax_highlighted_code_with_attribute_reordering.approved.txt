<!DOCTYPE html>
<article>
  <section>
    <pre id="code" class="code contain"><span></span><span id="line-1"><span class="k">def</span><span class="w"> </span><span class="nf">nth_root</span><span class="p">(</span><span class="n">radicand</span><span class="p">):</span>
</span>    <span id="line-2" class="fragment initial-highlighted-initial" data-fragment-group-id="single-return"><span class="k">return</span> <span class="n">radicand</span>
</span></pre>
  </section>
</article>
